### PR TITLE
fix: R2Score returns NaN instead of 1.0 for a perfect prediction on zero-variance data

### PR DESCRIPTION
**Pull request by gaurav-gandhi-2411**
## Summary

`R2Score.result()` guards against `-inf` (via `ops.isinf`) but not `NaN`. When a class has zero `y_true` variance (`total == 0`) and the prediction is exact (`total_mse == 0`), `total_mse / total` is `0/0 = NaN`, which passes the `isinf` guard unguarded and poisons the aggregated score under `uniform_average`/`variance_weighted_average`.

Keras documents `R2Score` as scikit-learn-equivalent (`class_aggregation` docstring: *"Equivalent to `multioutput` argument in Scikit-Learn"*; `test_r2_sklearn_comparison` validates numeric parity against sklearn for standard inputs), but has no coverage for the zero-variance case, where the current implementation diverges from sklearn's `force_finite` behavior.

sklearn's `r2_score` (`force_finite=True`, the default) treats a zero numerator as a perfect prediction and returns `1.0` regardless of the denominator — quoting the actual source, `sklearn/metrics/_regression.py::_assemble_r2_explained_variance`:

```python
nonzero_denominator = denominator != 0
nonzero_numerator = numerator != 0
# Default = Zero Numerator = perfect predictions. Set to 1.0
# (note: even if denominator is zero, thus avoiding NaN scores)
output_scores = xp.ones([n_outputs], device=device, dtype=dtype)
valid_score = nonzero_denominator & nonzero_numerator
output_scores[valid_score] = 1 - (numerator[valid_score] / denominator[valid_score])
# Non-zero Numerator and Zero Denominator:
# arbitrary set to 0.0 to avoid -inf scores
output_scores[nonzero_numerator & ~nonzero_denominator] = 0.0
```

This PR makes `R2Score` match: `NaN` (0/0, zero variance + perfect prediction) → `1.0`. The existing `-inf` (nonzero/0, zero variance + imperfect prediction) → `0.0` case is untouched and was already correct.

## Testing

Added `test_r2_zero_variance_perfect_prediction` (expects `1.0`; fails on `master` with a NaN-location mismatch before this fix) and `test_r2_zero_variance_imperfect_prediction` (expects `0.0`; locks in the already-correct existing behavior for the neighboring case).

Full `keras/src/metrics/` suite, all three backends:
- `KERAS_BACKEND=jax`: 407 passed
- `KERAS_BACKEND=torch`: 407 passed
- `KERAS_BACKEND=tensorflow`: 407 passed

## Contributor Agreement

- [x] I am a human, and not a bot.
- [x] I will be responsible for responding to review comments in a timely manner.
- [x] I will work with the maintainers to push this PR forward until submission.

## AI-Assisted Contribution Disclosure

Per the AI-Assisted Contribution Policy: I used an AI coding agent (Claude Code)
in preparing this PR. It was used to locate the defect, draft the fix and the two
regression tests, and run the verification described above. I reviewed the change,
verified the sklearn behaviour against `sklearn/metrics/_regression.py` myself,
and confirmed the test results across the jax, torch, and tensorflow backends.
I understand the change and will own responses to review.

### CODE REVIEW

File: keras/src/metrics/regression_metrics.py [hunk 1]
```diff
@@ -534,6 +534,11 @@ def result(self):
         mean = self.sum / self.count
         total = self.squared_sum - self.sum * mean
         raw_scores = 1 - (self.total_mse / total)
+        # total == 0 (zero-variance y_true) and total_mse == 0 (perfect
+        # prediction) is 0/0 == NaN; sklearn treats a zero numerator as a
+        # perfect prediction and scores it 1.0 regardless of the
+        # denominator (sklearn.metrics._regression, `force_finite=True`).
+        raw_scores = ops.where(ops.isnan(raw_scores), 1.0, raw_scores)
         raw_scores = ops.where(ops.isinf(raw_scores), 0.0, raw_scores)
```
[the remark this answers is not in the document]
Comment (gaurav-gandhi-2411) [reply keras/src/metrics/regression_metrics.py]: Thanks — the masking concern is real: I confirmed with a numeric check that `isnan(raw_scores)` would silently turn a NaN from a diverged model (normal variance, NaN prediction) into a spurious 1.0. Applied your suggestion as proposed and added `test_r2_nan_total_mse_propagates` to lock in that a NaN under normal variance surfaces as NaN, not 1.0.

**Comment by gaurav-gandhi-2411** [comment]
This fixes `R2Score.result()` returning NaN instead of 1.0 for a perfect prediction on zero-variance targets, matching sklearn's `force_finite` behavior. The state-variable-based approach suggested in the initial review was applied, CI is green across all backends, and the added tests pass. This is tagged `stat:awaiting keras-eng` — is there anything blocking a keras engineer from taking a look?

File: keras/src/metrics/regression_metrics.py [hunk 2]
```diff
@@ -533,7 +533,18 @@ def update_state(self, y_true, y_pred, sample_weight=None):
...
+        # on properties of the computed ratio: a NaN in total_mse can also
+        # come from unrelated numerical instability (e.g. exploding
+        # gradients), and checking isnan(raw_scores) can't tell that case
+        # apart from the deliberate 0/0 of a zero-variance perfect
+        # prediction. It would silently report a perfect score instead of
+        # surfacing the NaN.
+        safe_total = ops.where(ops.equal(total, 0.0), 1.0, total)
+        raw_scores = 1.0 - (self.total_mse / safe_total)
+        raw_scores = ops.where(ops.equal(total, 0.0), 0.0, raw_scores)
+        raw_scores = ops.where(ops.equal(self.total_mse, 0.0), 1.0, raw_scores)
```
Comment (hertschuh) [inline keras/src/metrics/regression_metrics.py:547]: I believe this whole thing can just be replaced with:

```python
raw_scores = 1.0 - ops.divide_no_nan(self.total_mse, total)
```

**Comment by hertschuh** [comment]
Thanks for addressing this!

File: keras/src/metrics/regression_metrics.py (hunk 2 above)
Comment (gaurav-gandhi-2411) [reply keras/src/metrics/regression_metrics.py:547]: `divide_no_nan` does handle the perfect-prediction zero-variance case correctly (0/0 → 0 → raw_scores 1.0), but it regresses the imperfect-prediction one: swapping it in, `test_r2_zero_variance_imperfect_prediction` fails on every backend (numpy/tensorflow/jax/torch, identical failure each time) with `ACTUAL: array(1., dtype=float32)` vs `DESIRED: array(0.)`, since a nonzero `total_mse` over a zero `total` also collapses to 0 under `divide_no_nan`. That's structural, not a backend quirk: `divide_no_nan` only ever inspects the denominator, so it can't distinguish that case from the genuine 0/0 that should map to 1.0 — the two differ only in the numerator, which the one-liner never looks at. I'll add a comment above the explicit branches instead, since readability is probably the real concern behind the suggestion — happy to also rename `safe_total` if that reads clearer.

**Review by hertschuh — APPROVED** [review]
Understood. Thanks!
